File: customer-service/customer_service/database/repositories/customer_repository.py
"""
Customer Repository for MAA Banking System
Handles all customer-related database operations
"""

import logging
import uuid
from typing import Optional, List, Dict, Any
from datetime import datetime, date
from ..  import execute_query, execute_mutation

logger = logging.getLogger(__name__)


class CustomerRepository:
    """Repository class for customer operations"""
    
    def get_customer_by_id(self, customer_id: str) -> Optional[Dict[str, Any]]:
        """
        Get customer by customer_id with full profile including accounts and appointment history
        
        Args:
            customer_id: The customer's ID
            
        Returns:
            Complete customer profile dictionary or None if not found
        """
        try:
            # Get customer basic info
            customer = execute_query(
                "SELECT * FROM customers WHERE customer_id = %s",
                (customer_id,),
                fetch_one=True
            )
            
            if not customer:
                return None
            
            # Get bank accounts
            accounts = execute_query(
                "SELECT * FROM bank_accounts WHERE customer_id = %s ORDER BY created_at",
                (customer['id'],)
            )
            
            # Get appointment history with branch info
            appointments = execute_query("""
                SELECT a.*, b.name as branch_name, b.location_code 
                FROM appointments a 
                JOIN branches b ON a.branch_id = b.id 
                WHERE a.customer_id = %s 
                ORDER BY a.appointment_date DESC, a.appointment_time DESC
            """, (customer['id'],))
            
            # Format the response to match the existing structure
            return {
                "contact_id": str(customer['id']),  # Keep for compatibility
                "customer_id": customer['customer_id'],
                "customer_first_name": customer['first_name'],
                "customer_last_name": customer['last_name'],
                "email": customer['email'],
                "phone_number": customer['phone_number'],
                "customer_start_date": customer['customer_start_date'].isoformat() if customer['customer_start_date'] else None,
                "years_as_customer": customer['years_as_customer'] or 0,
                "billing_address": customer['billing_address'] or {},
                "accounts": [
                    {
                        "account_id": acc['account_id'],
                        "account_type": acc['account_type'],
                        "balance": float(acc['balance']),
                        "status": acc['status']
                    }
                    for acc in (accounts or [])
                ],
                "appointment_history": [
                    {
                        "appointment_id": apt['appointment_id'],
                        "reason": apt['reason'],
                        "date": apt['appointment_date'].isoformat() if apt['appointment_date'] else None,
                        "time": str(apt['appointment_time'])[:5] if apt['appointment_time'] else None,  # Format as HH:MM
                        "location": apt['location_code'],
                        "status": apt['status'],
                        "banker_name": apt['banker_name']
                    }
                    for apt in (appointments or [])
                ],
                "preferred_branch": (customer['bank_profile'] or {}).get('preferred_branch', 'Manhattan'),
                "communication_preferences": customer['communication_preferences'] or {
                    "email": True,
                    "sms": True, 
                    "push_notifications": False
                },
                "bank_profile": customer['bank_profile'] or {},
                "current_appointment": None  # Will be set if there's a pending appointment
            }
            
        except Exception as e:
            logger.error(f"Error getting customer {customer_id}: {e}")
            return None
    
    def create_customer(self, customer_data: Dict[str, Any]) -> Optional[str]:
        """
        Create a new customer
        
        Args:
            customer_data: Customer information dictionary
            
        Returns:
            Customer ID if successful, None otherwise
        """
        try:
            result = execute_mutation("""
                INSERT INTO customers (
                    customer_id, first_name, last_name, email, phone_number,
                    customer_start_date, years_as_customer, billing_address,
                    communication_preferences, bank_profile
                ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
                RETURNING customer_id
            """, (
                customer_data['customer_id'],
                customer_data['first_name'],
                customer_data['last_name'],
                customer_data['email'],
                customer_data.get('phone_number'),
                customer_data.get('customer_start_date'),
                customer_data.get('years_as_customer', 0),
                customer_data.get('billing_address'),
                customer_data.get('communication_preferences'),
                customer_data.get('bank_profile')
            ), return_id=True)
            
            return result['customer_id'] if result else None
            
        except Exception as e:
            logger.error(f"Error creating customer: {e}")
            return None
    
    def update_customer(self, customer_id: str, updates: Dict[str, Any]) -> bool:
        """
        Update customer information
        
        Args:
            customer_id: Customer ID to update
            updates: Dictionary of fields to update
            
        Returns:
            True if successful, False otherwise
        """
        try:
            # Build dynamic update query
            set_clauses = []
            values = []
            
            for field, value in updates.items():
                set_clauses.append(f"{field} = %s")
                values.append(value)
            
            if not set_clauses:
                return False
            
            values.append(customer_id)
            
            rowcount = execute_mutation(f"""
                UPDATE customers 
                SET {', '.join(set_clauses)}, updated_at = CURRENT_TIMESTAMP
                WHERE customer_id = %s
            """, tuple(values))
            
            return rowcount > 0
            
        except Exception as e:
            logger.error(f"Error updating customer {customer_id}: {e}")
            return False


# Global instance
customer_repository = CustomerRepository()

File: MAA/backend/database/repositories/CustomerRepository.js
/**
 * Customer Repository
 * Handles all customer-related database operations
 */

import { query, withTransaction } from '../connection.js';

export class CustomerRepository {
  /**
   * Find customer by customer ID
   * @param {string} customerId - Customer ID
   * @returns {Promise<Object|null>} Customer object or null
   */
  async findByCustomerId(customerId) {
    const result = await query(
      'SELECT * FROM customers WHERE customer_id = $1',
      [customerId]
    );
    return result.rows[0] || null;
  }

  /**
   * Find customer by email
   * @param {string} email - Customer email
   * @returns {Promise<Object|null>} Customer object or null
   */
  async findByEmail(email) {
    const result = await query(
      'SELECT * FROM customers WHERE email = $1',
      [email]
    );
    return result.rows[0] || null;
  }

  /**
   * Get customer with all related data (accounts, appointments)
   * @param {string} customerId - Customer ID
   * @returns {Promise<Object|null>} Complete customer profile
   */
  async getCustomerProfile(customerId) {
    const customer = await this.findByCustomerId(customerId);
    if (!customer) return null;

    // Get bank accounts
    const accountsResult = await query(
      'SELECT * FROM bank_accounts WHERE customer_id = $1 ORDER BY created_at',
      [customer.id]
    );

    // Get appointment history
    const appointmentsResult = await query(`
      SELECT a.*, b.name as branch_name, b.location_code 
      FROM appointments a 
      JOIN branches b ON a.branch_id = b.id 
      WHERE a.customer_id = $1 
      ORDER BY a.appointment_date DESC, a.appointment_time DESC
    `, [customer.id]);

    return {
      contact_id: customer.id, // For compatibility with existing code
      customer_id: customer.customer_id,
      customer_first_name: customer.first_name,
      customer_last_name: customer.last_name,
      email: customer.email,
      phone_number: customer.phone_number,
      customer_start_date: customer.customer_start_date,
      years_as_customer: customer.years_as_customer,
      billing_address: customer.billing_address,
      accounts: accountsResult.rows.map(account => ({
        account_id: account.account_id,
        account_type: account.account_type,
        balance: parseFloat(account.balance),
        status: account.status
      })),
      appointment_history: appointmentsResult.rows.map(apt => ({
        appointment_id: apt.appointment_id,
        reason: apt.reason,
        date: apt.appointment_date,
        time: apt.appointment_time.substring(0, 5), // Format HH:MM
        location: apt.location_code,
        status: apt.status,
        banker_name: apt.banker_name
      })),
      preferred_branch: customer.bank_profile?.preferred_branch || 'Manhattan',
      communication_preferences: customer.communication_preferences,
      bank_profile: customer.bank_profile,
      current_appointment: null // Will be populated if there's a pending appointment
    };
  }

  /**
   * Create a new customer
   * @param {Object} customerData - Customer data
   * @returns {Promise<Object>} Created customer
   */
  async create(customerData) {
    const {
      customer_id,
      first_name,
      last_name,
      email,
      phone_number,
      customer_start_date,
      years_as_customer,
      billing_address,
      communication_preferences,
      bank_profile
    } = customerData;

    const result = await query(`
      INSERT INTO customers (
        customer_id, first_name, last_name, email, phone_number,
        customer_start_date, years_as_customer, billing_address,
        communication_preferences, bank_profile
      ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)
      RETURNING *
    `, [
      customer_id, first_name, last_name, email, phone_number,
      customer_start_date, years_as_customer, 
      JSON.stringify(billing_address),
      JSON.stringify(communication_preferences),
      JSON.stringify(bank_profile)
    ]);

    return result.rows[0];
  }

  /**
   * Update customer information
   * @param {string} customerId - Customer ID
   * @param {Object} updates - Fields to update
   * @returns {Promise<Object>} Updated customer
   */
  async update(customerId, updates) {
    const customer = await this.findByCustomerId(customerId);
    if (!customer) throw new Error('Customer not found');

    const setClause = [];
    const values = [];
    let paramIndex = 1;

    Object.keys(updates).forEach(key => {
      if (key === 'billing_address' || key === 'communication_preferences' || key === 'bank_profile') {
        setClause.push(`${key} = $${paramIndex}`);
        values.push(JSON.stringify(updates[key]));
      } else {
        setClause.push(`${key} = $${paramIndex}`);
        values.push(updates[key]);
      }
      paramIndex++;
    });

    values.push(customer.id);

    const result = await query(`
      UPDATE customers 
      SET ${setClause.join(', ')}, updated_at = CURRENT_TIMESTAMP
      WHERE id = $${paramIndex}
      RETURNING *
    `, values);

    return result.rows[0];
  }

  /**
   * Delete a customer and all related data
   * @param {string} customerId - Customer ID
   * @returns {Promise<boolean>} Success status
   */
  async delete(customerId) {
    const customer = await this.findByCustomerId(customerId);
    if (!customer) return false;

    await query('DELETE FROM customers WHERE id = $1', [customer.id]);
    return true;
  }

  /**
   * Search customers by name or email
   * @param {string} searchTerm - Search term
   * @returns {Promise<Array>} Array of customers
   */
  async search(searchTerm) {
    const result = await query(`
      SELECT customer_id, first_name, last_name, email, phone_number
      FROM customers 
      WHERE 
        first_name ILIKE $1 OR 
        last_name ILIKE $1 OR 
        email ILIKE $1
      ORDER BY last_name, first_name
      LIMIT 50
    `, [`%${searchTerm}%`]);

    return result.rows;
  }
}

export default new CustomerRepository();
